File: src/priv/bajdcc/LALR1/semantic/test/TestSemantic3.java
package priv.bajdcc.LALR1.semantic.test;

import priv.bajdcc.LALR1.semantic.Semantic;
import priv.bajdcc.LALR1.semantic.token.ISemanticAnalyzer;
import priv.bajdcc.LALR1.syntax.handler.SyntaxException;
import priv.bajdcc.util.lexer.error.RegexException;
import priv.bajdcc.util.lexer.token.TokenType;

@SuppressWarnings("unused")
public class TestSemantic3 {

	public static void main(String[] args) {
		// System.out.println("Z -> `a`<,> | B | [`a` `b` Z B]");
		try {
			// Scanner scanner = new Scanner(System.in);
			// String expr = "( i )";
			String expr = "a a a a";
			Semantic semantic = new Semantic(expr);
			semantic.addTerminal("a", TokenType.ID, "a");
			semantic.addNonTerminal("START");
			semantic.addNonTerminal("Z");
			semantic.addErrorHandler("sample", null);
			ISemanticAnalyzer handleValue = (indexed, query, recorder) -> 1;
			ISemanticAnalyzer handleCopy = (indexed, query, recorder) -> indexed.get(0).object;
			ISemanticAnalyzer handleRec = (indexed, query, recorder) -> {
                int lop = Integer
                        .parseInt(indexed.get(0).object.toString());
                return lop + 1;
            };
			// syntax.infer("E -> T `PLUS`<+> E | T `MINUS`<-> E | T");
			// syntax.infer("T -> F `TIMES`<*> T | F `DIVIDE`</> T | F");
			// syntax.infer("F -> `LPA`<(> E `RPA`<)>  | `SYMBOL`<i>");
			semantic.infer(handleCopy, "START -> Z[0]");
			semantic.infer(handleRec, "Z -> Z[0] @a[1]");
			semantic.infer(handleValue, "Z -> @a[0]");
			semantic.initialize("START");
			System.out.println(semantic.toString());
			System.out.println(semantic.getNGAString());
			System.out.println(semantic.getNPAString());
			System.out.println(semantic.getInst());
			System.out.println(semantic.getTrackerError());
			System.out.println(semantic.getTokenList());
			System.out.println(semantic.getObject());
			// scanner.close();
		} catch (RegexException e) {
			System.err.println(e.getPosition() + "," + e.getMessage());
			e.printStackTrace();
		} catch (SyntaxException e) {
			System.err.println(e.getPosition() + "," + e.getMessage() + " "
					+ e.getInfo());
			e.printStackTrace();
		}
	}
}

File: src/main/java/com/bajdcc/LALR1/semantic/test/TestSemantic3.kt
package com.bajdcc.LALR1.semantic.test

import com.bajdcc.LALR1.grammar.semantic.ISemanticAction
import com.bajdcc.LALR1.grammar.semantic.ISemanticAnalyzer
import com.bajdcc.LALR1.grammar.semantic.ISemanticRecorder
import com.bajdcc.LALR1.grammar.symbol.IManageSymbol
import com.bajdcc.LALR1.grammar.symbol.IQuerySymbol
import com.bajdcc.LALR1.semantic.Semantic
import com.bajdcc.LALR1.semantic.token.IIndexedData
import com.bajdcc.LALR1.semantic.token.IRandomAccessOfTokens
import com.bajdcc.LALR1.syntax.handler.SyntaxException
import com.bajdcc.util.lexer.error.RegexException
import com.bajdcc.util.lexer.token.TokenType

object TestSemantic3 {

    @JvmStatic
    fun main(args: Array<String>) {
        // System.out.println("Z -> `a`<,> | B | [`a` `b` Z B]");
        try {
            // Scanner scanner = new Scanner(System.in);
            // String expr = "( i )";
            val expr = "a a a a"
            val semantic = Semantic(expr)
            semantic.addTerminal("a", TokenType.ID, "a")
            semantic.addTerminal("b", TokenType.EPSILON, null)
            semantic.addNonTerminal("START")
            semantic.addNonTerminal("Z")
            val handleValue = object : ISemanticAnalyzer {
                override fun handle(indexed: IIndexedData, query: IQuerySymbol, recorder: ISemanticRecorder): Any {
                    return 1
                }
            }
            val handleCopy = object : ISemanticAnalyzer {
                override fun handle(indexed: IIndexedData, query: IQuerySymbol, recorder: ISemanticRecorder): Any {
                    return indexed.get(0).obj!!
                }
            }
            val handleRec = object : ISemanticAnalyzer {
                override fun handle(indexed: IIndexedData, query: IQuerySymbol, recorder: ISemanticRecorder): Any {
                    val lop = Integer
                            .parseInt(indexed.get(0).obj!!.toString())
                    return lop + 1
                }
            }
            val action = object : ISemanticAction {
                override fun handle(indexed: IIndexedData, manage: IManageSymbol, access: IRandomAccessOfTokens, recorder: ISemanticRecorder) {
                    println("ok")
                }
            }
            // syntax.infer("E -> T `PLUS`<+> E | T `MINUS`<-> E | T");
            // syntax.infer("T -> F `TIMES`<*> T | F `DIVIDE`</> T | F");
            // syntax.infer("F -> `LPA`<(> E `RPA`<)>  | `SYMBOL`<i>");
            semantic.addActionHandler("Action", action)
            semantic.infer(handleCopy, "START -> Z[0]")
            semantic.infer(handleRec, "Z -> @a[1] Z[0]#Action# @b")
            semantic.infer(handleValue, "Z -> @a[0]")
            semantic.initialize("START")
            println(semantic.toString())
            println(semantic.ngaString)
            println(semantic.npaString)
            semantic.parse()
            println(semantic.inst)
            println(semantic.trackerError)
            println(semantic.tokenList)
            println(semantic.getObject())
            // scanner.close();
        } catch (e: RegexException) {
            System.err.println(e.position.toString() + "," + e.message)
            e.printStackTrace()
        } catch (e: SyntaxException) {
            System.err.println(e.position.toString() + "," + e.message + " "
                    + e.info)
            e.printStackTrace()
        }

    }
}
